File: expand_image.py
#!/usr/bin/env python3
"""
Image Aspect Ratio Expander
Expands images from 16:1 to 1.91:1 aspect ratio by adding padding on the sides.
"""

from PIL import Image
import argparse
import os
from pathlib import Path


def expand_image(
    input_path: str,
    output_path: str = None,
    target_ratio: float = 1.91,
    padding_color: str = "black",
    source_ratio: float = 16.0
) -> str:
    """
    Expand an image from source aspect ratio to target aspect ratio.

    Args:
        input_path: Path to the input image
        output_path: Path to save the output image (optional)
        target_ratio: Target aspect ratio (width:height), default 1.91:1
        padding_color: Color for padding ('black' or 'white'), default 'black'
        source_ratio: Source aspect ratio (width:height), default 16:1

    Returns:
        Path to the output image
    """
    # Open the image
    img = Image.open(input_path)
    original_width, original_height = img.size

    print(f"Original image size: {original_width}x{original_height}")
    print(f"Original aspect ratio: {original_width/original_height:.2f}:1")

    # Calculate the new dimensions
    # For 16:1 -> 1.91:1, we need to add padding on left and right
    # The height stays the same, width changes based on target ratio
    new_height = original_height
    new_width = int(new_height * target_ratio)

    print(f"Target aspect ratio: {target_ratio}:1")
    print(f"New image size: {new_width}x{new_height}")

    # Determine padding color
    if padding_color.lower() == "white":
        bg_color = (255, 255, 255)
        if img.mode == "RGBA":
            bg_color = (255, 255, 255, 255)
    else:  # black
        bg_color = (0, 0, 0)
        if img.mode == "RGBA":
            bg_color = (0, 0, 0, 255)

    # Create new image with padding
    new_img = Image.new(img.mode, (new_width, new_height), bg_color)

    # Calculate position to paste the original image (centered)
    x_offset = (new_width - original_width) // 2
    y_offset = 0

    # Paste the original image onto the new background
    new_img.paste(img, (x_offset, y_offset))

    # Generate output path if not provided
    if output_path is None:
        input_file = Path(input_path)
        output_path = str(input_file.parent / f"{input_file.stem}_expanded{input_file.suffix}")

    # Save the new image
    new_img.save(output_path)
    print(f"Saved expanded image to: {output_path}")

    return output_path


def main():
    parser = argparse.ArgumentParser(
        description="Expand images from 16:1 to 1.91:1 aspect ratio with padding"
    )
    parser.add_argument(
        "input",
        help="Input image path"
    )
    parser.add_argument(
        "-o", "--output",
        help="Output image path (default: adds '_expanded' suffix to input filename)",
        default=None
    )
    parser.add_argument(
        "-c", "--color",
        choices=["black", "white"],
        default="black",
        help="Padding color (default: black)"
    )
    parser.add_argument(
        "-t", "--target-ratio",
        type=float,
        default=1.91,
        help="Target aspect ratio width:height (default: 1.91)"
    )
    parser.add_argument(
        "-s", "--source-ratio",
        type=float,
        default=16.0,
        help="Source aspect ratio width:height (default: 16.0)"
    )

    args = parser.parse_args()

    # Check if input file exists
    if not os.path.exists(args.input):
        print(f"Error: Input file '{args.input}' does not exist")
        return 1

    try:
        expand_image(
            args.input,
            args.output,
            args.target_ratio,
            args.color,
            args.source_ratio
        )
        return 0
    except Exception as e:
        print(f"Error: {e}")
        return 1


if __name__ == "__main__":
    exit(main())

File: src/expand-image.ts
#!/usr/bin/env node

/**
 * Image Aspect Ratio Expander
 * Expands images from 16:1 to 1.91:1 aspect ratio by adding padding on the sides.
 */

import sharp from 'sharp';
import { program } from 'commander';
import { promises as fs } from 'fs';
import path from 'path';

interface ExpandOptions {
  targetRatio?: number;
  paddingColor?: 'black' | 'white';
  sourceRatio?: number;
  paddingStrategy?: 'solid' | 'blur' | 'edge-extend';
  blurAmount?: number;
}

/**
 * Expand an image from source aspect ratio to target aspect ratio.
 */
async function expandImage(
  inputPath: string,
  outputPath?: string,
  options: ExpandOptions = {}
): Promise<string> {
  const {
    targetRatio = 1.91,
    paddingColor = 'black',
    sourceRatio = 16.0,
    paddingStrategy = 'solid',
    blurAmount = 10
  } = options;

  // Check if input file exists
  try {
    await fs.access(inputPath);
  } catch (error) {
    throw new Error(`Input file '${inputPath}' does not exist`);
  }

  // Get image metadata
  const image = sharp(inputPath);
  const metadata = await image.metadata();

  if (!metadata.width || !metadata.height) {
    throw new Error('Unable to read image dimensions');
  }

  const originalWidth = metadata.width;
  const originalHeight = metadata.height;
  const originalRatio = originalWidth / originalHeight;

  console.log(`Original image size: ${originalWidth}x${originalHeight}`);
  console.log(`Original aspect ratio: ${originalRatio.toFixed(2)}:1`);

  // Calculate new dimensions
  const newHeight = originalHeight;
  const newWidth = Math.round(newHeight * targetRatio);

  console.log(`Target aspect ratio: ${targetRatio}:1`);
  console.log(`New image size: ${newWidth}x${newHeight}`);
  console.log(`Padding strategy: ${paddingStrategy}`);

  // Calculate position to center the original image
  const xOffset = Math.floor((newWidth - originalWidth) / 2);
  const paddingWidth = xOffset;

  let expandedImage: Buffer;

  if (paddingStrategy === 'edge-extend') {
    // Use Sharp's extend with edge repeat
    expandedImage = await image
      .extend({
        left: paddingWidth,
        right: paddingWidth,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .png()
      .toBuffer();

    // Since extend doesn't have a true edge-repeat, we'll manually create it
    const inputBuffer = await image.png().toBuffer();

    // Extract left edge strip (1 pixel wide)
    const leftEdge = await sharp(inputBuffer)
      .extract({ left: 0, top: 0, width: 1, height: originalHeight })
      .resize(paddingWidth, originalHeight, { kernel: 'nearest' })
      .toBuffer();

    // Extract right edge strip (1 pixel wide)
    const rightEdge = await sharp(inputBuffer)
      .extract({ left: originalWidth - 1, top: 0, width: 1, height: originalHeight })
      .resize(paddingWidth, originalHeight, { kernel: 'nearest' })
      .toBuffer();

    // Composite everything together
    expandedImage = await sharp({
      create: {
        width: newWidth,
        height: newHeight,
        channels: metadata.hasAlpha ? 4 : 3,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      }
    })
      .composite([
        { input: leftEdge, top: 0, left: 0 },
        { input: inputBuffer, top: 0, left: paddingWidth },
        { input: rightEdge, top: 0, left: paddingWidth + originalWidth }
      ])
      .png()
      .toBuffer();

  } else if (paddingStrategy === 'blur') {
    // Create blurred edge padding
    const inputBuffer = await image.png().toBuffer();

    // Extract left edge strip (20 pixels wide for better blur effect)
    const edgeWidth = Math.min(20, Math.floor(originalWidth * 0.1));

    const leftEdge = await sharp(inputBuffer)
      .extract({ left: 0, top: 0, width: edgeWidth, height: originalHeight })
      .blur(blurAmount)
      .resize(paddingWidth, originalHeight, { fit: 'fill' })
      .toBuffer();

    // Extract right edge strip
    const rightEdge = await sharp(inputBuffer)
      .extract({
        left: originalWidth - edgeWidth,
        top: 0,
        width: edgeWidth,
        height: originalHeight
      })
      .blur(blurAmount)
      .resize(paddingWidth, originalHeight, { fit: 'fill' })
      .toBuffer();

    // Composite everything together
    expandedImage = await sharp({
      create: {
        width: newWidth,
        height: newHeight,
        channels: metadata.hasAlpha ? 4 : 3,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      }
    })
      .composite([
        { input: leftEdge, top: 0, left: 0 },
        { input: inputBuffer, top: 0, left: paddingWidth },
        { input: rightEdge, top: 0, left: paddingWidth + originalWidth }
      ])
      .png()
      .toBuffer();

  } else {
    // Solid color padding (original behavior)
    const backgroundColor = paddingColor === 'white'
      ? { r: 255, g: 255, b: 255, alpha: 1 }
      : { r: 0, g: 0, b: 0, alpha: 1 };

    const inputBuffer = await image.png().toBuffer();

    expandedImage = await sharp({
      create: {
        width: newWidth,
        height: newHeight,
        channels: metadata.hasAlpha ? 4 : 3,
        background: backgroundColor
      }
    })
      .composite([
        {
          input: inputBuffer,
          top: 0,
          left: xOffset
        }
      ])
      .png()
      .toBuffer();
  }

  // Generate output path if not provided
  if (!outputPath) {
    const parsedPath = path.parse(inputPath);
    outputPath = path.join(
      parsedPath.dir,
      `${parsedPath.name}_expanded${parsedPath.ext}`
    );
  }

  // Save the expanded image
  await sharp(expandedImage).toFile(outputPath);
  console.log(`Saved expanded image to: ${outputPath}`);

  return outputPath;
}

// CLI interface
if (require.main === module) {
  program
    .name('expand-image')
    .description('Expand images from 16:1 to 1.91:1 aspect ratio with intelligent padding')
    .argument('<input>', 'Input image path')
    .option('-o, --output <path>', 'Output image path')
    .option(
      '-c, --color <color>',
      'Padding color for solid strategy (black or white)',
      'black'
    )
    .option(
      '-p, --padding-strategy <strategy>',
      'Padding strategy: solid, blur, or edge-extend',
      'solid'
    )
    .option(
      '-b, --blur-amount <amount>',
      'Blur amount for blur strategy (default: 10)',
      '10'
    )
    .option(
      '-t, --target-ratio <ratio>',
      'Target aspect ratio width:height',
      '1.91'
    )
    .option(
      '-s, --source-ratio <ratio>',
      'Source aspect ratio width:height',
      '16.0'
    )
    .action(async (input, options) => {
      try {
        const color = options.color as 'black' | 'white';
        const strategy = options.paddingStrategy as 'solid' | 'blur' | 'edge-extend';

        if (color !== 'black' && color !== 'white') {
          console.error('Error: Color must be either "black" or "white"');
          process.exit(1);
        }

        if (!['solid', 'blur', 'edge-extend'].includes(strategy)) {
          console.error('Error: Padding strategy must be "solid", "blur", or "edge-extend"');
          process.exit(1);
        }

        await expandImage(input, options.output, {
          targetRatio: parseFloat(options.targetRatio),
          paddingColor: color,
          sourceRatio: parseFloat(options.sourceRatio),
          paddingStrategy: strategy,
          blurAmount: parseFloat(options.blurAmount)
        });

        process.exit(0);
      } catch (error) {
        console.error(`Error: ${(error as Error).message}`);
        process.exit(1);
      }
    });

  program.parse();
}

export { expandImage, ExpandOptions };
